<h1>UK ROAD ACCIDENT DATA ANALYSIS</h1>
<h2>INCLUSIVE YEAR 2019 - 2022</h2>
<h3>Analyst: John Benedict Lorredo</h3>

<h2>DATA PREPARATION</h2>
<h3>Transfer Data set to folder</h3>

<h2>Importing Libraries</h2>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import f_oneway

<h1>DATA FRAME</h1>

In [2]:
accident = pd.read_csv('datasets\\accident_data.csv')

In [3]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [4]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

<h1>Converting Date Field into DATE TIME type</h1>

In [5]:
accident['Accident Date']

0           5/6/2019
1           2/7/2019
2         26-08-2019
3         16-08-2019
4           3/9/2019
             ...    
660674    18-02-2022
660675    21-02-2022
660676    23-02-2022
660677    23-02-2022
660678    28-02-2022
Name: Accident Date, Length: 660679, dtype: object

In [6]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [7]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [8]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unknown road condition')
accident['Road_Type'] = accident['Road_Type'].fillna('unaccounted')
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unaccounted')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])


In [9]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [10]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2              unaccounted               Urban            unaccounted   
3       Single carriageway               Urban     Fine no high winds   
4              unaccounted               Urban            unaccounted   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [11]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [12]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')

In [13]:
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')

In [14]:
accident['District Area'] = accident['District Area'].astype('category')

In [15]:
accident['Number_of_Casualties'] = accident['Number_of_Casualties'].astype('category')

In [16]:
accident['Number_of_Vehicles'] = accident['Number_of_Vehicles'].astype('category')

In [17]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')

In [18]:
accident['Road_Type'] = accident['Road_Type'].astype('category')

In [19]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')

In [20]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')

In [21]:
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [22]:
accident['Latitude'] = accident['Latitude'].astype('category')
accident['Longitude'] = accident['Longitude'].astype('category')

In [23]:
accident.dtypes

Index                        object
Accident_Severity          category
Accident Date                object
Latitude                   category
Light_Conditions           category
District Area              category
Longitude                  category
Number_of_Casualties       category
Number_of_Vehicles         category
Road_Surface_Conditions    category
Road_Type                  category
Urban_or_Rural_Area        category
Weather_Conditions         category
Vehicle_Type               category
dtype: object

In [24]:
accident['Accident Date'] = accident['Accident Date'].astype('str')
accident.dtypes
accident['Accident Date'] = accident['Accident Date'].str.strip()
accident['Accident Date'] = accident['Accident Date'].str.replace('/','-')

In [25]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

In [26]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [27]:
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                         category
Light_Conditions                 category
District Area                    category
Longitude                        category
Number_of_Casualties             category
Number_of_Vehicles               category
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

In [28]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayOfWeek'] = accident['Accident Date'].dt.dayofweek

In [29]:
latitude_longitude_cor = accident['Latitude'].corr(accident['Longitude'])

In [30]:
latitude_longitude_cor

np.float64(-0.3981137948101014)

In [31]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area Longitude  \
0             Darkness - lights lit  Kensington and Chelsea -0.209082   
1                          Daylight  Kensington and Chelsea -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea -0.210215   
3                          Daylight  Kensington and Chelsea -0.202731   
4             Darkness - lights lit  Kensington and Chelsea -0.192487   
...                             ...                     ...       ...   
660674                     Daylight                Highland -3.467828   
660675       Darkness - no lighting                Highland -3.809281   
660676                     Daylight                Highland -3.862727   
660677       Darkness - no lighting                Highland -3.823997   
660678                     Daylight                Highland -3.895673   

       Number_of_Casualties Number_of_Vehicles Road_Surface_Conditions  \
0                         1                  2                     Dry   
1                         1                  2             Wet or damp   
2                         1                  3                     Dry   
3                         1                  4                     Dry   
4                         1                  2                     Dry   
...                     ...                ...                     ...   
660674                    2                  1                     Dry   
660675                    1                  1            Frost or ice   
660676                    1                  3            Frost or ice   
660677                    1                  2             Wet or damp   
660678                    1                  1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2              unaccounted               Urban            unaccounted   
3       Single carriageway               Urban     Fine no high winds   
4              unaccounted               Urban            unaccounted   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  Year  Month  Day  DayOfWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...                                      ...   ...    ...  ...        ...  
660674                  

<H1>1</H1>

In [32]:
accident.groupby(['Road_Type', 'Accident_Severity']).size().unstack()

C:\Users\Meloujean\AppData\Local\Temp\ipykernel_15312\1685114215.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  accident.groupby(['Road_Type', 'Accident_Severity']).size().unstack()


Accident_Severity   Fatal  Serious  Slight
Road_Type                                 
Dual carriageway     1815    11746   85863
One way street         95     1655   11809
Roundabout            142     3665   40185
Single carriageway   6527    70059  415557
Slip road              49      611    6381
unaccounted            33      481    4006

<H1>INSIGHT 1</H1>
<H2>The highest number of accidents occur on single carriageways, with 415,557 slight, 70,059 serious, and 6,527 fatal accidents.</H2>

<h1>2</h1>

In [37]:
yearly_accidents = accident['Year'].value_counts()

In [38]:
yearly_accidents

Year
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

<h1>INSIGHT 2</h1>
<H2>Accidents have decreased over the years, from 182,115 in 2019 to 144,419 in 2022.</H2>

<h1>3</h1>

In [40]:
accident.groupby(['Weather_Conditions', 'Accident_Severity']).size().unstack()

C:\Users\Meloujean\AppData\Local\Temp\ipykernel_15312\4032887760.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  accident.groupby(['Weather_Conditions', 'Accident_Severity']).size().unstack()


Accident_Severity      Fatal  Serious  Slight
Weather_Conditions                           
Fine + high winds        175     1245    7134
Fine no high winds      7100    72046  441739
Fog or mist               82      483    2963
Other                    165     1801   15184
Raining + high winds     145     1261    8209
Raining no high winds    848     9468   69380
Snowing + high winds       3      109     773
Snowing no high winds     36      565    5637
unaccounted              107     1239   12782

<h1>INSIGHT 3</h1>
<H2>Most accidents occur in clear weather. "Fine, no high winds" accounts for 7,100 fatal, 72,046 serious, and 441,739 slight accidents.</H2>

<h1>4 and 5</h1>

In [46]:
accident.groupby(['Month', 'Accident_Severity']).size().unstack()

C:\Users\Meloujean\AppData\Local\Temp\ipykernel_15312\2989858049.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  accident.groupby(['Month', 'Accident_Severity']).size().unstack()


Accident_Severity  Fatal  Serious  Slight
Month                                    
1                    740     6681   45451
2                    617     6246   42628
3                    681     7118   46287
4                    697     7336   43711
5                    760     7721   47871
6                    673     7916   47892
7                    753     7863   48829
8                    801     7532   45580
9                    723     7987   47745
10                   798     7939   50843
11                   747     7568   52109
12                   671     6310   44855

<h1>INSIGHT 4</h1>
<H2>November has the most accidents (60,424), while February has the fewest (49,491).</H2>

<h1>INSIGHT 5</h1>
<H2>August has the highest number of fatal accidents (801).</H2>

<h1>No. 6</h1>

In [47]:
accident.groupby(['Road_Surface_Conditions', 'Accident_Severity']).size().unstack()

C:\Users\Meloujean\AppData\Local\Temp\ipykernel_15312\2115508688.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  accident.groupby(['Road_Surface_Conditions', 'Accident_Severity']).size().unstack()


Accident_Severity        Fatal  Serious  Slight
Road_Surface_Conditions                        
Dry                       5788    61638  380395
Flood over 3cm. deep        23      152     842
Frost or ice               193     2007   16317
Snow                        35      565    5290
Wet or damp               2620    23785  160303
unknown road condition       2       70     654

<h1>INSIGHT NO. 6</h1>
<H2>The majority of accidents occur on dry roads, with a total of 448,821 incidents</H2>

<h1>No. 7</h1>

In [49]:
vehicle_counts = accident['Vehicle_Type'].value_counts()

In [50]:
vehicle_counts

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h1>INSIGHT NO. 7</h1>
<H2>Cars are the most frequently involved vehicle type in accidents, with a total of 497,992 cases</H2>

<h1>8 and 9</h1>

In [62]:
accident.groupby(['Urban_or_Rural_Area', 'Accident_Severity']).size().unstack()

C:\Users\Meloujean\AppData\Local\Temp\ipykernel_15312\285106878.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  accident.groupby(['Urban_or_Rural_Area', 'Accident_Severity']).size().unstack()


Accident_Severity    Fatal  Serious  Slight
Urban_or_Rural_Area                        
Rural                 5601    37312  196077
Unallocated              0        1      10
Urban                 3060    50904  367714

<h1>INSIGHT 8</h1>
<H2>Urban areas have the highest number of recorded accidents (420,778), with the majority classified as slight (367,714), followed by serious (50,904) and fatal (3,060)</H2>

<h1>INSIGHT No. 9</h1>
<h2>Rural areas account for 239,990 accidents, with slight (196,077), serious (37,312), and fatal (5,601) cases.</h2>

<h1>10, 11 and 12</h1>

In [65]:
accident.groupby(['Light_Conditions', 'Accident_Severity']).size().unstack()

C:\Users\Meloujean\AppData\Local\Temp\ipykernel_15312\712378712.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  accident.groupby(['Light_Conditions', 'Accident_Severity']).size().unstack()


Accident_Severity            Fatal  Serious  Slight
Light_Conditions                                   
Darkness - lighting unknown     68      794    5622
Darkness - lights lit         1860    19130  108345
Darkness - lights unlit         45      360    2138
Darkness - no lighting        1612     7174   28651
Daylight                      5076    60759  419045

<h1>Insight 10</h1>
<h2>The majority of accidents (69.79%) occur in daylight, with the majority classified as slight 419,045(86.32%), followed by serious 60,759 (12.51%) and fatal 5,076 (1.05%). </h2>

<h1>Insight 11</h1>
<h2>Accidents are more frequent in lit areas during darkness, with 129,395 cases occurring where streetlights are on, compared to 36,100 cases in areas with no lighting</h2>

<h1>INSIGHT NO. 12</h1>
<h2>Fatal accidents are most common in daylight, with 5,076 cases, followed by areas where lit at night 1,860 cases.</h2>

<h1>No. 13, 14, and 15</h1>

In [67]:
accident.groupby(['Year', 'Accident_Severity']).size().unstack()

C:\Users\Meloujean\AppData\Local\Temp\ipykernel_15312\2138974727.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  accident.groupby(['Year', 'Accident_Severity']).size().unstack()


Accident_Severity  Fatal  Serious  Slight
Year                                     
2019                2714    24322  155079
2020                2341    23121  145129
2021                2057    21997  139500
2022                1549    18777  124093

<h1>INSIGHT 13</h1>
<h2>The total number of accidents decreased each year, from 182,035 in 2019 to 144,439 in 2022, indicating a continuous decline</h2>

<h1>INSIGHT 14</h1>
<H2>Fatal accidents also declined over time, from 2,714 in 2019 to 1,549 in 2022</H2>

<H1>INSIGHT 15</H1>
<H2>The most significant decrease in total accidents happened between 2021 and 2022, with a decline of 17,114 cases, the highest drop in four years.</H2>

<h1>No. 16</h1>

In [69]:
accident['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h1>Insight No. 16</h1>
<h2>In 2022, about 65.8% of accidents in 2022 occurred in urban areas, while 34.9% in rural areas.</h2>

<h1>No. 17, 18, 19, 20</h1>

In [ ]:
accident[['Accident Date', 'DayOfWeek']].head(10)

In [ ]:
accident.groupby(['DayOfWeek', 'Accident_Severity']).size().unstack()

In [ ]:
72680, 94550, 99558, 99511, 97900, 107178, 89302

<h1>INSIGHT NO. 17</h1>
<h2>Saturday (Day 5) has the highest number of accidents, with 91,852 slight and 14,000 serious cases, making it the most accident-prone day of the week.</h2>

<h1>INSIGHT NO. 18</h1>
<h2>Sunday (Day 6) records the highest number of fatal accidents (1,522 cases), while Thursdays (Day 3) have the lowest (1,097)</h2> 

<h1>INSIGHT NO. 19</h1>
<h2>Monday (Day 0) has the lowest number of total accidents, </h2>

<h1>INSIGHT NO. 20</h1>
<h2>Wednesday (99558) and Saturday (107,178) together account for a significant portion of weekly accidents.</h2>